### PR TITLE
feat!: multiple buckets / providers support (#48)

### PR DESCRIPTION
**Pull request by hbollon**
As the title shown, I've added the multiple buckets / providers support into Terraboard.
It's now possible to set multiples S3/MinIO buckets or to mix providers (to have AWS, Gitlab and GCP simultaneously for example), Terraboard will fetch them all on db refresh.

However: 
- To use this functionality, the only configuration method usable is through Yaml config file. Env and flags are still usable for mono bucket configuration (maybe we can use them for multiple providers but it's not appropriated imo)
- Atm, Terraboard is dependent of **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** env variables for AWS provider, which prevents being able to have several backend AWS or others compatible providers like MinIO (we can still have several buckets under the same backend). We could consider making these values configurable via the yaml file however (@raphink what do you think?).

I tested that with MinIO/AWS but not with others providers since I don't have any account for them.
The `terraboard-dev-multi-providers` service in the `test/` docker-compose is configured to use two MinIO bucker (just replace `terraboard-dev` by it in the compose file)
I also updated the ReadMe with some instructions / advices.

Should meets the needs of #48 :smile:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,6 +90,9 @@ It currently supports several remote state backend providers:
 - [Terraform Cloud (remote)](https://www.terraform.io/docs/backends/types/remote.html)
 - [GitLab](https://docs.gitlab.com/ee/user/infrastructure/terraform_state.html)
 
+With the upcoming **v1.2.0** update, Terraboard will be now able to handle multiple buckets/providers configuration! 🥳
```
Comment (raphink) [inline README.md:93]: That's a nice addition. We'll just need to remember to remove it when we release 😄

File: README.md [hunk 2]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
 Terraboard currently supports configuration in three different ways:
 
-1. Environment variables
-2. CLI parameters
-3. Configuration file (YAML). A configuration file example can be found in the root directory of this repository.
+1. Environment variables **(only usable for mono provider configuration)**
+2. CLI parameters **(only usable for mono provider configuration)**
+3. Configuration file (YAML). A configuration file example can be found in the root directory of this repository and in the `test/` subdirectory.
+
+**Important: all flags/environment variables related to the providers settings aren't compatible with multi-provider configuration! For that, you must use the Yaml config file to be able to configure multiples buckets/providers.**
```
Comment (raphink) [inline README.md:153]:
```suggestion
**Important: all flags/environment variables related to the providers settings aren't compatible with multi-provider configuration! Instead, you must use the YAML config file to be able to configure multiples buckets/providers.**
```

File: README.md [hunk 3]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
 
 The precedence of configurations is as described below.
 
+### Multiple buckets/providers
+
+In order to be able to link to Terraboard multiples buckets or even providers, you must use the Yaml configuration method.
```
Comment (raphink) [inline README.md:159]:
```suggestion
In order for Terraboard to import states from multiples buckets or even providers, you must use the YAML configuration method.
```

File: README.md [hunk 4]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+
+- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to a valid Yaml config file.
```
Comment (raphink) [inline README.md:161]:
```suggestion
- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to point to a valid YAML config file.
```

File: README.md [hunk 5]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+
+- In the Yaml file, specify your desired providers configuration. For example with two S3 buckets:
```
Comment (raphink) [inline README.md:163]:
```suggestion
- In the YAML file, specify your desired providers configuration. For example for two S3 buckets:
```

File: README.md [hunk 6]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+        - .tfstate
+
+  - endpoint: http://minio:9000/
+    region: eu-west-1
+    s3:
+      bucket: test-bucket2
+      force-path-style: true
+      file-extension: 
+        - .tfstate
+
```
Comment (raphink) [inline README.md:182]:
```suggestion
```

File: README.md (hunk 5 above)
Comment (raphink) [reply README.md:163]: Don't we need to specify the no versioning and no locks options in this example since it uses minio?

File: README.md [hunk 7]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+```
+
+In the case of AWS, don't forget to set **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
```
Comment (raphink) [inline README.md:185]:
```suggestion
In the case of AWS, don't forget to set the **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
```

File: README.md [hunk 8]
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+
+That's it! Terraboard will now fetch these two buckets on DB refresh. You can also mix providers like AWS and Gitlab or anything else.
+You can find a ready-to-use docker example with two *MinIO* buckets in the `test/` sub-folder (just swipe the two terraboard services in the docker-compose file).
```
Comment (raphink) [inline README.md:188]:
```suggestion
You can find a ready-to-use Docker example with two *MinIO* buckets in the `test/` sub-folder (just swipe the two terraboard services in the docker-compose file). 
```

Comment (raphink) [reply README.md:188]: What do you mean by "swipe the services"?

File: api/api.go [hunk 9]
```diff
@@ -147,15 +147,21 @@ func StateCompare(w http.ResponseWriter, r *http.Request, d *db.Database) {
...
 
 // GetLocks returns information on locked States
-func GetLocks(w http.ResponseWriter, _ *http.Request, sp state.Provider) {
+func GetLocks(w http.ResponseWriter, _ *http.Request, sps []state.Provider) {
 	w.Header().Set("Access-Control-Allow-Origin", "*")
-	locks, err := sp.GetLocks()
-	if err != nil {
-		JSONError(w, "Failed to get locks", err)
-		return
+	allLocks := make(map[string]state.LockInfo)
```
Comment (raphink) [inline api/api.go:152]: What is the point of using a map rather than an array?

File: config/config.go [hunk 10]
```diff
@@ -119,14 +119,44 @@ func (c *Config) LoadConfigFromYaml() *Config {
...
 	if _, err := parser.Parse(); err != nil {
 		fmt.Printf("Failed to parse flags: %s", err)
 		os.Exit(1)
 	}
+}
+
+// LoadConfig loads the config from flags & environment
+func LoadConfig(version string) *Config {
+	c := initDefaultConfig()
+	parseStructFlagsAndEnv(&c)
```
Comment (raphink) [inline config/config.go:159]: Why is this necessary?

File: main.go [hunk 11]
```diff
@@ -64,56 +64,61 @@ func isKnownStateVersion(statesVersions map[string][]string, versionID, path str
...
-		states, err := sp.GetStates()
-		if err != nil {
-			log.WithFields(log.Fields{
-				"error": err,
-			}).Error("Failed to retrieve states. Retrying in 1 minute.")
-			time.Sleep(interval)
-			continue
-		}
+		log.Debugf("Total providers: %d\n", len(sps))
+		for i, sp := range sps {
```
Comment (raphink) [inline main.go:74]: Looping inside `refreshDB()` means each provider will be parsed one a time. Maybe we could parallelize providers instead by looping ouside the call to `refreshDB()` in `main()` and keep `refreshDB()` the way it is.

File: README.md (hunk 8 above)
Comment (hbollon) [reply README.md:188]: > What do you mean by "swipe the services"?

I added in the docker-compose of `/test`, a Terraboard container configured to use a YAML config file and two MinIO buckets.
However, I commented it in favor of the Terraboard lambda container to avoid having two applications simultaneously.

```yaml
services:
  terraboard-dev:
    build:
      context: ../
      dockerfile: ./Dockerfile
    environment:
      AWS_ACCESS_KEY_ID: root
      AWS_SECRET_ACCESS_KEY: mypassword
      AWS_BUCKET: test-bucket
      AWS_REGION: eu-west-1
      AWS_ENDPOINT: http://minio:9000/
      AWS_FORCE_PATH_STYLE: "true"
      TERRABOARD_LOG_LEVEL: debug
      TERRABOARD_NO_LOCKS: "true"
      TERRABOARD_NO_VERSIONING: "true"
      DB_PASSWORD: mypassword
      DB_SSLMODE: disable
      GODEBUG: netdns=go
    depends_on:
      db:
        condition: service_healthy
      minio:
        condition: service_started
    volumes:
      - ../static:/static:ro
    ports:
      - "8080:8080"

  # terraboard-dev-multi-providers:
  #   build:
  #     context: ../
  #     dockerfile: ./Dockerfile
  #   environment:
  #     AWS_ACCESS_KEY_ID: root
  #     AWS_SECRET_ACCESS_KEY: mypassword
  #     DB_SSLMODE: disable
  #     CONFIG_FILE: config/config.yml
  #     GODEBUG: netdns=go
  #   depends_on:
  #     db:
  #       condition: service_healthy
  #     minio:
  #       condition: service_started
  #   volumes:
  #     - ../static:/static:ro
  #     - ./config.yml:/config/config.yml:ro
  #   ports:
  #     - "8081:8080"
```

I'm not really happy with that, maybe we can split the test sub-folder into multiple config folders each with a docker-compose :thinking:

File: config/config.go (hunk 10 above)
Comment (hbollon) [reply config/config.go:159]: It's for keeping the env / flags configuration possible with single provider Terraboard.
Indeed, now we have providers array in the `config` struct:
```go
AWS []AWSConfig `group:"AWS Options" yaml:"aws"`

TFE []TFEConfig `group:"Terraform Enterprise Options" yaml:"tfe"`

GCP []GCPConfig `group:"Google Cloud Platform Options" yaml:"gcp"`

Gitlab []GitlabConfig `group:"GitLab Options" yaml:"gitlab"`
```

And so they don't have any elements in them on `Config` instanciation which prevent from using env variables or flags.
`c := initDefaultConfig()` add an empty struct in each of them and parse go-flags stuff in them. If they're still empty or invalid they will be ignored but with that process we can still use Terraboard with others configuration methods as before (although they are incompatible with multi buckets).

`parseStructFlagsAndEnv(interface{})` is just a DRY function for go-flags parsing

File: main.go (hunk 11 above)
Comment (hbollon) [reply main.go:74]: Great idea that I should have thought of! 
This will allow to not interrupt all the fetching processes if one of them has an error and reduce the time that the refresh takes
I will do that

File: README.md (hunk 8 above)
Comment (raphink) [reply README.md:188]: Ah, so you mean "swap" then, not "swipe" :-)

Comment (hbollon) [reply README.md:188]: Yes :laughing:

Comment (hbollon) [reply README.md:188]: Yes :laughing:

**Comment by hbollon** [comment]
I've done the following changes: 
- RefreshDB is now parallelized 
- I added mutex lock security on sensible db operations (FirstOrCreate principally, which lead to errors without locks)
- I split test directory into multiples sub-folder containing different configurations 
- I fixed all ReadMe typo issues

File: README.md [hunk 12]
```diff
@@ -142,12 +146,51 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
 
 The precedence of configurations is as described below.
 
+### Multiple buckets/providers
+
+In order for Terraboard to import states from multiples buckets or even providers, you must use the YAML configuration method.
+
+- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to point to a valid YAML config file.
+
+- In the YAML file, specify your desired providers configuration. For example with two MinIO buckets (using the AWS provider with compatible mode):
```
Comment (raphink) [inline README.md:163]:
```suggestion
In order for Terraboard to import states from multiples buckets or even providers, you must use the YAML configuration method:

- Set the `CONFIG_FILE` environment variable or the `-c`/`--config-file` flag to point to a valid YAML config file.
- In the YAML file, specify your desired providers configuration. For example with two MinIO buckets (using the AWS provider with compatible mode):
```

File: README.md [hunk 13]
```diff
@@ -142,12 +146,51 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+  - endpoint: http://minio:9000/
+    region: eu-west-1
+    s3:
+      bucket: test-bucket2
+      force-path-style: true
+      file-extension: 
+        - .tfstate
+```
+
+In the case of AWS, don't forget to set the **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
```
Comment (raphink) [inline README.md:189]: I'd use tt instead of bold for code
```suggestion
In the case of AWS, don't forget to set the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables.
```

File: db/db.go [hunk 14]
```diff
@@ -79,7 +81,7 @@ func Init(config config.DBConfig, debug bool) *Database {
 		db.Config.Logger.LogMode(logger.Info)
 	}
 
-	d := &Database{db}
+	d := &Database{DB: db}
```
Comment (raphink) [inline db/db.go:84]: Not sure this change is necessary (though it doesn't hurt)

File: db/db.go [hunk 15]
```diff
@@ -124,11 +126,13 @@ func (db *Database) stateS3toDB(sf *statefile.File, path string, versionID strin
 	// Check if the associated lineage is already present in lineages table
 	// If so, it recovers its ID otherwise it inserts it at the same time as the state
 	var lineage types.Lineage
+	db.lock.Lock()
 	err = db.FirstOrCreate(&lineage, types.Lineage{Value: sf.Lineage}).Error
 	if err != nil || lineage.ID == 0 {
-		log.Error("Unknown error in stateS3toDB during lineage finding")
+		log.Errorf("Unknown error in stateS3toDB during lineage finding: %v", err)
```
Comment (raphink) [inline db/db.go:132]: It could be interesting to use a structured log here (with fields).

File: db/db.go (hunk 14 above)
Comment (hbollon) [reply db/db.go:84]: I had _InvalidStructLit_ error from gopls without it since I added `lock sync.Mutex` field

File: db/db.go (hunk 15 above)
Comment (hbollon) [reply db/db.go]: Done :+1:

**Comment by Shocktrooper** [comment]
> As the title shown, I've added the multiple buckets / providers support into Terraboard.
> It's now possible to set multiples S3/MinIO buckets or to mix providers (to have AWS, Gitlab and GCP simultaneously for example), Terraboard will fetch them all on db refresh.
> 
> However:
> 
> * To use this functionality, the only configuration method usable is through Yaml config file. Env and flags are still usable for mono bucket configuration (maybe we can use them for multiple providers but it's not appropriated imo)
> * Atm, Terraboard is dependent of **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** env variables for AWS provider, which prevents being able to have several backend AWS or others compatible providers like MinIO (we can still have several buckets under the same backend). We could consider making these values configurable via the yaml file however (@raphink what do you think?).
> 
> I tested that with MinIO/AWS but not with others providers since I don't have any account for them.
> The `terraboard-dev-multi-providers` service in the `test/` docker-compose is configured to use two MinIO bucker (just replace `terraboard-dev` by it in the compose file)
> I also updated the ReadMe with some instructions / advices.
> 
> Should meets the needs of #48 😄

With **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** being needed does this prevent multiple cross account S3 buckets from being selected? Would they all have to be in the same account and is there work to migrate over to the AWS SDK or something that can automatically handle credentials automatically instead of relying on hardcoded environment variables

**Comment by hbollon** [comment]
> With **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** being needed does this prevent multiple cross account S3 buckets from being selected?

Actually, with this PR, yes all AWS buckets must be on the same account.

> is there work to migrate over to the AWS SDK or something that can automatically handle credentials automatically instead of relying on hardcoded environment variables

This is exactly what I proposed to @raphink 😃

**Comment by hbollon** [comment]
I've added:
- AWS access key and secret access key config fields and aws sdk appropriate usage to support multiple AWS accounts configuration
- AWS Session Token support (should do the trick for #181)
- The ability to directly configure multiple buckets for the same AWS account under the "S3" field of the YAML file which is now an objects array

I also updated unit test and docker testing env to match with new config (I also add a second minio container in the multiple-minio-buckets test env)
Like previously, I made sure that the config via environment variables or by flags is always possible for the instances of Terraboard with a single provider

File: README.md (hunk 1 above)
Comment (raphink) [reply README.md:93]:
```suggestion
With the upcoming **v2.0.0** update, Terraboard will be now able to handle multiple buckets/providers configuration! 🥳
```

File: config/yaml.go [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+	raw := rawGitlabConfig{
+		Address: "https://gitlab.com",
+	}
+	if err := unmarshal(&raw); err != nil {
+		return err
+	}
+
+	*s = GitlabConfig(raw)
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (hbollon) [reply config/yaml.go:33]: Yes, without that we don't have default values for Yaml config method

File: state/aws.go [hunk 17]
```diff
@@ -31,39 +31,45 @@ type AWS struct {
 }
 
 // NewAWS creates an AWS object
-func NewAWS(c *config.Config) AWS {
-	sess := session.Must(session.NewSession())
-
-	awsConfig := aws_sdk.NewConfig()
+func NewAWS(c *config.Config) []*AWS {
```
[the remark this answers is not in the document]
Comment (hbollon) [reply state/aws.go:35]: To avoid the iteration in `NewAWS()`?

**Comment by Shocktrooper** [comment]
Is there since there was mention of aws sdk usage of credentials is there an example of how to tell terraboard to assume a role in another account or how to tell it to use an instance profile in the case of S3 buckets

File: README.md [hunk 18]
```diff
@@ -142,12 +146,50 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
 
 Terraboard currently supports configuration in three different ways:
 
-1. Environment variables
-2. CLI parameters
-3. Configuration file (YAML). A configuration file example can be found in the root directory of this repository.
+1. Environment variables **(only usable for mono provider configuration)**
```
Comment (raphink) [inline README.md:149]:
```suggestion
1. Environment variables **(only usable for single provider configuration)**
```

File: README.md [hunk 19]
```diff
@@ -142,12 +146,50 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
...
+2. CLI parameters **(only usable for mono provider configuration)**
```
Comment (raphink) [inline README.md:150]:
```suggestion
2. CLI parameters **(only usable for single provider configuration)**
```

File: state/aws.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (raphink) [reply state/aws.go:35]: Yes, and also because generally I'd expect an instantiation method to return a single instance.

[the remark this answers is not in the document]
Comment (hbollon) [reply state/aws.go:35]: Ok so I will do that for AWS but also for all others providers
Each one will have a `NewProvider()` function wich return a single instance and a `NewProviderCollection()` one which will iterate and call the other one. (ofc in functions names _**Provider**_ will be replaced by his name)

File: state/aws.go [hunk 20]
```diff
@@ -31,39 +32,59 @@ type AWS struct {
...
-
-	awsConfig := aws_sdk.NewConfig()
-
-	if len(c.AWS.APPRoleArn) > 0 {
-		log.Debugf("Using %s role", c.AWS.APPRoleArn)
-		creds := stscreds.NewCredentials(sess, c.AWS.APPRoleArn, func(p *stscreds.AssumeRoleProvider) {
-			if c.AWS.ExternalID != "" {
-				p.ExternalID = aws_sdk.String(c.AWS.ExternalID)
+func NewAWS(aws config.AWSConfig, bucket config.S3BucketConfig, noLocks, noVersioning bool) *AWS {
+	if bucket.Bucket != "" {
```
Comment (raphink) [inline state/aws.go:36]: Could you shortcut this case instead?

```go
if bucket.Bucket == "" {
  return nil
}
```

File: state/gcp.go [hunk 21]
```diff
@@ -24,33 +24,53 @@ type GCP struct {
...
 // NewGCP creates an GCP object
-func NewGCP(c *config.Config) (*GCP, error) {
+func NewGCP(gcp config.GCPConfig) (*GCP, error) {
 	ctx := context.Background()
+
 	var client *storage.Client
+	var gcpInstance *GCP
 	var err error
-	if c.GCP.GCPSAKey != "" {
+	if gcp.GCSBuckets != nil {
```
Comment (raphink) [inline state/gcp.go:33]: Same here, I think the bucket logic could be short circuited

File: state/gitlab.go [hunk 22]
```diff
@@ -18,10 +18,25 @@ type Gitlab struct {
 }
 
 // NewGitlab creates a new Gitlab object
-func NewGitlab(c *config.Config) *Gitlab {
-	return &Gitlab{
-		Client: gitlab.NewClient(c.Gitlab.Address, c.Gitlab.Token),
+func NewGitlab(gl config.GitlabConfig) *Gitlab {
+	var instance *Gitlab
+	if gl.Token != "" {
```
Comment (raphink) [inline state/gitlab.go:23]: Same here, short circuit the empty token

File: state/state.go [hunk 23]
```diff
@@ -40,23 +40,53 @@ type Provider interface {
...
-func Configure(c *config.Config) (Provider, error) {
-	if len(c.TFE.Token) > 0 {
-		log.Info("Using Terraform Enterprise as the state/locks provider")
-		return NewTFE(c)
+func Configure(c *config.Config) ([]Provider, error) {
+	var providers []Provider
+	if len(c.TFE) > 0 {
+		objs, err := NewTFECollection(c)
+		if err != nil {
+			return []Provider{}, err
```
Comment (raphink) [inline state/state.go:48]: If you named the return values of the method, you might just use `return` here iirc

**Review by mcanevet — CHANGES_REQUESTED** [review]
I guess you should change one of your commits' commit message with an exclamation mark as it contains a breaking change.

File: state/state.go (hunk 23 above)
Comment (raphink) [reply state/state.go:48]: Note, if one provider fails, you get no providers at all at the moment.

File: state/tfe.go [hunk 24]
```diff
@@ -19,24 +19,42 @@ type TFE struct {
...
 
 // NewTFE creates a new TFE object
-func NewTFE(c *config.Config) (*TFE, error) {
-	config := &tfe.Config{
-		Address: c.TFE.Address,
-		Token:   c.TFE.Token,
-	}
+func NewTFE(tfeObj config.TFEConfig) (*TFE, error) {
+	var tfeInstance *TFE
+	if tfeObj.Token != "" {
```
Comment (raphink) [inline state/tfe.go:24]: Short circuit

File: state/state.go (hunk 23 above)
Comment (hbollon) [reply state/state.go:48]: > Note, if one provider fails, you get no providers at all at the moment.
Do you think it's better to just log an error and continue with the following ones?
Initially, I considered this error as fatal since it would lead to the non-inclusion of an entire provider and because, potentially, it could come from an error in the configuration file which could impact the rest.

Comment (raphink) [reply state/state.go:48]: Not sure. I say I'd prefer to see some data and get an error somewhere rather than nothing at all.

Comment (hbollon) [reply state/state.go:48]: > If you named the return values of the method, you might just use return here iirc

Indeed but in this case I think it's more simple like this since if I name the return values I can't use `:=` anymore with err which will force me to declare the temp variable `objs` before each time I use it. It's, imo, more complex and furthermore it raise an error on `gosimple` if I do it.
